### PR TITLE
feature/235-implement user registeration

### PR DESCRIPTION
**Pull request by Dharakkkk**
1. Implemented the UserService and UserServiceImpl:
   - Added methods for user registration, password hashing, email validation, and password validation.
   - Added necessary checks for existing email, password strength, and email format.

2. Updated UserRepository:
   - Added the existsByEmail method to check for existing users.

3. Created integration tests for the implemented functionality in UserServiceImplTest:
   - Tests for successful user registration, handling existing email, invalid email, and weak password.

4. Integrated user registration into AuthController:
   - Added a POST endpoint for user registration.

5. Added required dependencies to pom.xml:
   - Included the spring-security-crypto dependency for password hashing."

#### Checklist template for LEADERS

&#x2611; | &#x2610;

&#x2611; New code is covered with tests and coverage requirements are met.
&#x2611; Build is not failing.
&#x2611; All tests are passing.
&#x2610; All conflicts are resolved and the branch is rebased on top of the up-to-date parent branch.
&#x2611; All issue requirements are implemented.
&#x2610; At least two reviewers are assigned.
&#x2611; The Pull request is linked to the related issue.
&#x2611; Correct branch name. Convention [https://github.com/bycza-zagroda/track-expenses-app-documentation/blob/develop/githubwork/GITHUB_WORK.md]
&#x2611; Correct unit test name - suffix 'Test' in the end.
&#x2611; Correct integration test name - suffix 'IT' in the end.
&#x2611; Positive and negative test paths have been implemented.

### CODE REVIEW

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/AuthController.java [hunk 1]
```diff
@@ -20,6 +23,7 @@
 @RequiredArgsConstructor
 @Validated
 public class AuthController {
+private final UserService userService;
```
Comment (poneciak57) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/AuthController.java:26]: Code formating: Enter between class proprties and methods

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/repository/UserRepository.java [hunk 2]
```diff
@@ -4,4 +4,5 @@
 import pl.byczazagroda.trackexpensesappbackend.model.User;
 
 public interface UserRepository extends JpaRepository<User, Long> {
+    boolean existsByEmail(String email);
```
Comment (poneciak57) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/repository/UserRepository.java:7]: Code formating: Enter between class proprties and methods

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserService.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package pl.byczazagroda.trackexpensesappbackend.service;
+
+
+import pl.byczazagroda.trackexpensesappbackend.dto.AuthRegisterDTO;
+
+import javax.validation.Valid;
+
+public interface UserService {
+    void registerUser(@Valid AuthRegisterDTO authRegisterDTO);
```
Comment (poneciak57) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserService.java:9]: Code formating: Enter between class proprties and methods

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import pl.byczazagroda.trackexpensesappbackend.repository.UserRepository;
+
+import javax.validation.Valid;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@RequiredArgsConstructor
+@Service
+@Validated
+public class UserServiceImpl implements UserService{
```
Comment (poneciak57) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java:20]: Code formating: Enter between class proprties and methods format the whole class

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.any;
+import org.mockito.ArgumentCaptor;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class UserServiceImplTest {
+    @Mock
```
Comment (poneciak57) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImplTest.java:24]: Code formating: Enter between class proprties and methods format the whole class

**Review by poneciak57 — CHANGES_REQUESTED** [review]
Missing proper code formatting

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    private final UserRepository userRepository;
+    @Override
+    public void registerUser(@Valid AuthRegisterDTO authRegisterDTO) {
+        if (userRepository.existsByEmail(authRegisterDTO.email())) {
+            throw new IllegalArgumentException("A user with the same email already exists");
```
Comment (poneciak57) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java:25]: You should change the exception to the `AppRuntimeException` according to our error handling strategy u can refer to other services. It applies to whole service.

**Review by poneciak57 — CHANGES_REQUESTED** [review]
Requested changes

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/integration/UserServiceImplIT.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import pl.byczazagroda.trackexpensesappbackend.dto.AuthRegisterDTO;
+import pl.byczazagroda.trackexpensesappbackend.model.User;
+import pl.byczazagroda.trackexpensesappbackend.repository.UserRepository;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class UserServiceImplIT {
```
Comment (poneciak57) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/integration/UserServiceImplIT.java:18]: Should be `UserRegisterationIT`

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/integration/UserServiceImplIT.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    private UserRepository userRepository;
+
+    @BeforeEach
+    void setup() {
+        userRepository.deleteAll();
+    }
+
+    @Test
+    public void testRegisterUser() {
+        String url = "/api/auth/register";
```
Comment (poneciak57) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/integration/UserServiceImplIT.java:33]: url should be `private static final String` class field

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/integration/UserServiceImplIT.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Autowired
+    private UserRepository userRepository;
+
+    @BeforeEach
+    void setup() {
+        userRepository.deleteAll();
+    }
+
+    @Test
```
Comment (poneciak57) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/integration/UserServiceImplIT.java:31]: Tests should have description

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/integration/UserServiceImplIT.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    public void testRegisterUser() {
```
Comment (poneciak57) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/integration/UserServiceImplIT.java:32]: we have naming convention for tests like
`test<what your are testing>_when<case like "when user already exists">_then<what should happen>`

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImplTest.java (hunk 5 above)
Comment (poneciak57) [reply src/test/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImplTest.java:25]: still not resolved

**Review by poneciak57 — CHANGES_REQUESTED** [review]
Requested changes

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/integration/UserRegistrationIT.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.web.client.TestRestTemplate;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import pl.byczazagroda.trackexpensesappbackend.BaseIntegrationTestIT;
+import pl.byczazagroda.trackexpensesappbackend.dto.AuthRegisterDTO;
+import pl.byczazagroda.trackexpensesappbackend.model.User;
+import pl.byczazagroda.trackexpensesappbackend.repository.UserRepository;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (poneciak57) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/integration/UserRegistrationIT.java:16]: do not use asterix imports, change to static imports

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+import org.mockito.*;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+import pl.byczazagroda.trackexpensesappbackend.dto.AuthRegisterDTO;
+import pl.byczazagroda.trackexpensesappbackend.exception.AppRuntimeException;
+import pl.byczazagroda.trackexpensesappbackend.exception.ErrorCode;
+import pl.byczazagroda.trackexpensesappbackend.model.User;
+import pl.byczazagroda.trackexpensesappbackend.repository.UserRepository;
+
+import static org.junit.Assert.*;
```
Comment (poneciak57) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImplTest.java:16]: Do not use asterix imports, change to static imports

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    private UserServiceImpl userService;
+
+    @BeforeEach
+    public void setUp() {
+        Mockito.reset(userRepository);
+    }
+
+    @DisplayName("When password is valid, it should return a hashed password of correct length")
+    @Test
+    public void testHashPassword_WhenPasswordIsValid_ThenReturnHashed() {
```
Comment (poneciak57) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImplTest.java:51]: like in your integration tests When and Then should be lowercase so
from `testHashPassword_WhenPasswordIsValid_ThenReturnHashed`
to `testHashPassword_whenPasswordIsValid_thenReturnHashed`
applies to whole class

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
+package pl.byczazagroda.trackexpensesappbackend.service;
+
+import org.junit.Test;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.*;
```
Comment (poneciak57) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImplTest.java:7]: Do not use asterix imports, change to static imports

**Review by poneciak57 — CHANGES_REQUESTED** [review]
Asterix imports and tests methods names need to be fixed

**Review by poneciak57 — DISMISSED** [review]
Looks good to me

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    @DisplayName("When registering a user with too short password, an AppRuntimeException (U004) should be thrown")
+    @Test
+    public void testRegisterUser_whenPasswordIsTooShort_thenThrowException() {
+        AppRuntimeException exception = assertThrows(AppRuntimeException.class,
+                () -> userService.registerUser(REGISTER_DTO_TOO_SHORT_PASSWORD));
+        assertEquals(ErrorCode.U004.getBusinessMessage(), exception.getMessage());
+        verify(userRepository, never()).save(any(User.class));
+    }
```
Comment (dawciobiel) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImplTest.java:117]: Please add one empty line after this bracket

**Review by dawciobiel — CHANGES_REQUESTED** [review]
Request changes

**Review by dawciobiel — CHANGES_REQUESTED** [review]
changes required

**Review by dawciobiel — CHANGES_REQUESTED** [review]
Changes required

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+@RequiredArgsConstructor
+@Service
+@Validated
+public class UserServiceImpl implements UserService {
+
+    private final UserRepository userRepository;
+
+    @Override
+    public void registerUser(@Valid AuthRegisterDTO authRegisterDTO) {
+        if (userRepository.existsByEmail(authRegisterDTO.email())) {
```
Comment (dawciobiel) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java:27]: I think it's better to verify the e-mail address beforehand, and only then refer to the repository to check it.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+        user.setEmail(authRegisterDTO.email());
+        user.setPassword(hashedPassword);
+        user.setUserName(authRegisterDTO.username());
+        user.setUserStatus(status);
+
+        userRepository.save(user);
+    }
+
+    public void validateEmail(String email) {
+        String emailRegex = "^[A-Za-z0-9+_.-]+@(.+)$";
```
[the remark this answers is not in the document]
Comment (dawciobiel) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java:60]: Variables still have different names and serve the same purpose.
L:60 `emailRegex `
and L:68 `pattern`

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/integration/UserRegistrationIT.java [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+
+    @DisplayName("When trying to register a user that already exists, it should return BAD_REQUEST")
+    @Test
+    void testRegisterUser_whenUserAlreadyExists_thenShouldReturnErrorResponse() {
+        ResponseEntity<String> response = restTemplate
+                .postForEntity(REGISTER_USER_URL, REGISTER_DTO, String.class);
+
+        final User existingUser = User.builder()
+                .email("user@server.com")
+                .userName("User Bolek")
```
Comment (dawciobiel) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/integration/UserRegistrationIT.java:70]: Here you got:
```java
.userName("User Bolek")
```
but on line L:26 you got another user name.
```java
    private static final AuthRegisterDTO REGISTER_DTO =
            new AuthRegisterDTO("user@server.com", "User123@", "User_Bolek");
```

**Review by dawciobiel — CHANGES_REQUESTED** [review]
Request changes

**Review by dawciobiel — APPROVED** [review]
Looks good to me

**Review by poneciak57 — APPROVED** [review]
Looks good to me

**Pull request merged**